### PR TITLE
new-download-fix

### PR DESCRIPTION
**Pull request by AnkurPrabhu**
Issue Link https://github.com/LibrePhotos/librephotos/issues/375  @derneuere
New Pr for  #1055 
what has been done : 

have created a job type in long_running_job for download
created two method one for making the zip file i.e zip_photos_task and second which call the zip function i.e create_download_job
created a get api for the same view class which checks the status of the running job
Pending:

checking file system storage before starting the zipping process
test file

### CODE REVIEW

**Comment by AnkurPrabhu** [comment]
@derneuere  the pre-commit hooks does not  work for me in local while commiting code  i run it manually with `pre-commit run --all-files`

**Comment by derneuere** [comment]
I think you just have to install with pip install pre-commit and install the dev dependencies and then it should complain on all commits :)

**Comment by derneuere** [comment]
Oh, and you need to do a pre-commit install once it's installed!

**Comment by derneuere** [comment]
You don't have to change the proxy.conf folder as protected_media and its subfolders get already rerouted :)
Yes, from my understanding, nginx handles download automagically and should use chunking or something similar.

**Comment by AnkurPrabhu** [comment]
hey i am a little confused here as in what are all the things that are pending for this issue. I have listed some below pls let me know if there are more 
pending work for this pr : 
- test file 
- nginx/rerouting
- front end pr for using the new downlaod api

**Comment by derneuere** [comment]
Just think feature first. We are still trying to fix the bug and handle large downloads.

Right now, we should see a new long-running job, when we click on download with new api. This starts zipping the file. When the job is done, indicated with the long-running job, we should then call api/media/zip/<filename>.zip, to start downloading.

There are still two issues, feature wise. The user wants the zip to get deleted after the download, because he wants to conserve space. For this, we have now two new cases:
 
1. We need to delete the zip when we successfully downloaded the file. 
2. If the user closes the window or quits the download process, we also need to handle the zip. We can either delete it with an expiry window, or if the download stops :)

**Comment by AnkurPrabhu** [comment]
i was thinking of making something like a cron job which schedules the delete_zip_from_local job one day after it has been created 
also you want me to put these two use cases in the same pr or a different one ?

**Comment by derneuere** [comment]
You can use this schedule function for it! Just schedule the job, when you create the zipping job https://django-q2.readthedocs.io/en/master/schedules.html

Yes this all belongs in the same PR as it is one feature :)

File: api/views/views.py [hunk 1]
```diff
@@ -586,3 +592,51 @@ def post(self, request, format=None):
...
+        response={"job_id": job_id,'url':filename}
+
+        return Response(data=response,status=200)
+        
+   
+    def get(self,request):
+        try:
+            job=LongRunningJob.objects.get(job_id=request.data["job_id"])
+            if job.finished:
+                zip_file_path =  os.path.join("/protected_media/all_zip_folder/", request.data["url"])
```
Comment (sickelap) [inline api/views/views.py]: use constant from the settings (MEDIA_ROOT)

Comment (AnkurPrabhu) [reply api/views/views.py]: done

**Comment by AnkurPrabhu** [comment]
@derneuere @sickelap  after using nginx its pretty fast and i loved it !! can you pls tell me how did you set nginx up for librephotos also the test file is remaining now can you pls guide me through it not sure how to go about it  also should i start working on the frontend pr ?

**Comment by sickelap** [comment]
Ideally, we should have a managed task that we could run as a cronjob. As a one of the tasks, we could do a cleanup for whatever we need to cleanup...
check [commands](https://github.com/LibrePhotos/librephotos/tree/dev/api/management/commands) for an inspiration

**Comment by AnkurPrabhu** [comment]
> Ideally, we should have a managed task that we could run as a cronjob. As a one of the tasks, we could do a cleanup for whatever we need to cleanup... check [commands](https://github.com/LibrePhotos/librephotos/tree/dev/api/management/commands) for an inspiration

so should i remove the delete function i schedule during creation ? or this cron job is a different one which admins can use ?

**Comment by AnkurPrabhu** [comment]
I have added a test file which tests 2 scenarios :-
- when we have enough storage 
- when we do not have sufficient storage

**Comment by AnkurPrabhu** [comment]
hey @derneuere can we have the label "HACKTOBERFEST-ACCEPTED" for  both of my prs #1057 and #778 as hacktober will end soon and i just want 2 prs to be accepted/approved/merged it would really mean a lot and would motivate me, thank-you

**Comment by derneuere** [comment]
Added the labels to the pull requests and to the ones, that were already merged :)

The nginx proxy gets configured in here: https://github.com/LibrePhotos/librephotos-docker/blob/main/proxy/nginx.conf

I don't think that you have to add another location, as the zip folder is a subfolder of protected_media.

You can either use cronjobs or the schedule function in django q2. Both mechanism can be used to solve the issue, that the zip gets deleted after a time period.

The scheduled async task or the cronjob should get added once a zip gets created. It als should handle it gracefully, if the zip is already missing. 

I would prefer the schedule function as the user can see this in the django-admin dashboard.

The testcases look good! If you can think of other cases, which could break this function, then add cases for it too :)

**Comment by AnkurPrabhu** [comment]
> Added the labels to the pull requests and to the ones, that were already merged :)

Hey thank-you soo much for this and all the time you have given for guiding me really ...means a lot <3


> The nginx proxy gets configured in here: https://github.com/LibrePhotos/librephotos-docker/blob/main/proxy/nginx.conf

  thankyou 

> I don't think that you have to add another location, as the zip folder is a subfolder of protected_media.

   i did not understand this i am using the protected_media constant 

> You can either use cronjobs or the schedule function in django q2. Both mechanism can be used to solve the issue, that the zip gets deleted after a time period.

  I have used schedule once the file is created in the file `api/all_tasks.py`

> The scheduled async task or the cronjob should get added once a zip gets created. It als should handle it gracefully, if the zip is already missing.

 i have addressed this in `delete_zip_file` function in the file `api/all_tasks.py`
> The testcases look good! If you can think of other cases, which could break this function, then add cases for it too :)

  sure

**Comment by AnkurPrabhu** [comment]
@derneuere @sickelap can we discuss what is exactly missing/remaining here was planning to close this soon

**Comment by derneuere** [comment]
- A frontend PR, which makes it possible to test this end to end
- A way that the zip gets removed after download
- Removal of the old download method

**Comment by AnkurPrabhu** [comment]
> * A way that the zip gets removed after download

i schedule the delete method once its created do you want to also delete it if it has been downloaded by the user ?

**Comment by derneuere** [comment]
Yes, either by triggering an endpoint or by some other mechanisms. If I download multiple albums or have multiple users, I do not want my hard drive to be filled up, if it is not necessary.

**Comment by sickelap** [comment]
One more thing. It seems that the archive will be accessible by anyone who know the link. Is this intentional? Should we have a check to validate that only an owner can download the archive? Just a thought...

**Comment by derneuere** [comment]
Hmm, true. The easiest fix is, to put the user id into the filename, when zipping and checking for it when downloading. To enhance the security a bit more, the download name should not be guessable. so something like a <uuid + user id>.zip should work fine.

**Comment by AnkurPrabhu** [comment]
> Yes, either by triggering an endpoint or by some other mechanisms. If I download multiple albums or have multiple users, I do not want my hard drive to be filled up, if it is not necessary.

i just saw that we are storing everything(images,videos) locally right ? so for multiple users everyone will have photos on their local disk then why do we have a download option why are we not just pointing the disk location of the zip file instead of sending it to the user through frontend ?

**Comment by derneuere** [comment]
In this case locally means a self-hosted system, which may or may not be the same system as the client. You can have users, which do not have direct access to the server like for example friends, which want to download the pictures from the last party. In that case a download function is still needed.

**Comment by AnkurPrabhu** [comment]
> In this case locally means a self-hosted system, which may or may not be the same system as the client. You can have users, which do not have direct access to the server like for example friends, which want to download the pictures from the last party. In that case a download function is still needed.

oh okay so if i use libre photos and want to share access with my friends my friends can download images but those images will be from my device storage ? so if any other user uploads something it will be stored in my device ? asking these because i really wanted to know more about the product

**Comment by derneuere** [comment]
It's similar to other photo management systems. The systems which hosts the docker container acts as a cloud for the other devices. Other devices can upload images to your cloud, if they have an account and can share them from your cloud.

LibrePhotos acts as a multi user system, which means it can have multiple accounts. The images will then just be placed in a subfolder for each user.

**Comment by sickelap** [comment]
My opinion LP is not a file sharing platform. As I understood the feature is to download your photos to local machine, not to create an archive for others to download. I agree with @derneuere and photo sharing feature within the LP should be enough for sharing.

**Comment by AnkurPrabhu** [comment]
@derneuere @sickelap i have created a delete api which we can call from front-end as soon as the download is successfull
also the changes for uuid+username has been added will start working on the front-end pr now

**Comment by sickelap** [comment]
I think relying on client to call the endpoint for cleaning up might not be the best idea. The client could fail for any reason. This could be done few ways:

- create a scheduler within an application using django-apshcheduler and add a task to cleanup files after certain amount of time
- create a managed task that would go through all registered jobs and execute them as they are configured. one of jobs could be cleanup files. then add it to crontab to run periodically,
- similar to previous but create a simple dedicated managed task with single purpose to cleanup the files.

IMO 3rd option is simplest to implement 

Please share any other ideas

**Comment by AnkurPrabhu** [comment]
> Please share any other ideas
 
i also have some ideas 
1) we could call my delete function once the client has called the download api i.e before redirecting it to nginx we could just schedule it for say like after 10 -20 mins before sending the response
2) we could use something called as webhooks which we can send from the client that they have received the download file so we can create/trigger an event .... i dont have much experience creating webhooks will need guidance if so
3) and third is the one i already implemented i.e once the file is downloaded we can just send a request so that the user has the file and it would be efficient because as soon as the user has it it would be deleted and the edge case where the user might get disconnected is still covered because we have already scheduled the delete task after 1 day which will delete the file safely if it exists

File: api/views/views.py [hunk 2]
```diff
@@ -586,3 +612,69 @@ def post(self, request, format=None):
...
+            else:
+                return Response(
+                    data={"status": "PENDING", "progress": job.result}, status=202
+                )
+        except BaseException as e:
+            logger.error(str(e))
+            return Response(status=404)
+
+
+class DeleteZipView(APIView):
```
Comment (derneuere) [inline api/views/views.py:641]: This can be part of ZipListPhotosView_V2, I think :)

File: api/views/views.py [hunk 3]
```diff
@@ -251,6 +253,7 @@ class MediaAccessView(APIView):
     permission_classes = (AllowAny,)
 
     def _get_protected_media_url(self, path, fname):
+        print(fname)
```
Comment (derneuere) [inline api/views/views.py]: I think you forgot to remove this :)

File: api/all_tasks.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+    except Exception as e:
+        util.logger.error("Error while converting files to zip: {}".format(e))
+
+    lrj.finished_at = datetime.now().replace(tzinfo=pytz.utc)
+    lrj.finished = True
+    lrj.save()
+    # scheduling a task to delete the zip file after a day
+    execution_time = timezone.now() + timezone.timedelta(days=1)
+    schedule("api.all_tasks.delete_zip_file", filename, next_run=execution_time)
```
Comment (derneuere) [inline api/all_tasks.py:71]: @sickelap The first point is already implemented. Right now, the zip file will get deleted after a day. The delete endpoint is the happy path, this clean up job is the edge case, where the user never downloads the zip.

**Comment by derneuere** [comment]
I think we thought of most edge cases and looks pretty good. Found some smaller things and it would be great if you could remove the old view. Next step is to create the frontend pr, so that we can test it end to end :)

File: api/views/views.py (hunk 3 above)
Comment (AnkurPrabhu) [reply api/views/views.py]: will do it

File: api/views/views.py (hunk 2 above)
Comment (AnkurPrabhu) [reply api/views/views.py:641]: oh okay cool then will implement the delete inside the zip view only

File: api/all_tasks.py (hunk 4 above)
Comment (AnkurPrabhu) [reply api/all_tasks.py:71]: will start with the front-end pr 🚀

File: api/views/views.py (hunk 2 above)
Comment (AnkurPrabhu) [reply api/views/views.py:641]: hey i can do this but then i need the file path so sending it through body wont be good for a  delete request moreover would have to edit the urls.py for make call to a function based view but then the pattern of all class based views would not continue so i was thinking of making it as a separate view class

File: api/all_tasks.py (hunk 4 above)
Comment (sickelap) [reply api/all_tasks.py:71]: yea, I have missed this one. I think this should be enough for automated cleanup.

**Comment by AnkurPrabhu** [comment]
@sickelap  @derneuere 
hey my plan for front-end implementation : 
- create a data for download so that it can be shared across components (redux store)
- there are 3 componenets download_component, download_tray, and download_pannel
- download pannel will be a left side menu panel 
 Screenshot 2023-11-03 at 8 46 02 PM

- download tray will open when clicked on download pannel where we can see individual tasks 
Screenshot 2023-11-03 at 8 47 18 PM

- create a download component which replaces the current downlod option it basically sends data to the component download_pannel(left sidemenu) which creates a tray object and when clicked we can see the download progress n stuff 
- once the download is completed it can have the same code where it make a window object and click on the link automatically 
- thing is i dont have a lot of idea about redux so can you tell me how to go about it 
- pls let me know if changes are needed

**Comment by sickelap** [comment]
I would recommend not over-complicate things and don't think about redux. Focus on the more simple download option first. When that's done you could think of other use case scenarios.

**Comment by sickelap** [comment]
and be sure that all checks are green ;)

**Comment by AnkurPrabhu** [comment]
> I would recommend not over-complicate things and don't think about redux. Focus on the more simple download option first. When that's done you could think of other use case scenarios.

okay got it

**Comment by derneuere** [comment]
Also, just use the notifications (https://v5.mantine.dev/core/notification/) to communicate, that zipping started or that the download is now starting. No need to make it work as a new component on the side.

**Comment by AnkurPrabhu** [comment]
Hey i had rebased my branch with dev and then forced pushed becuase it was still showing branch is behind  but the error that  my backend container is restarting continuously is because of  ImportError: cannot import name 'dynamo_export' from 'torch.onnx'  which was addded 2 days ago what should i do ?

**Comment by AnkurPrabhu** [comment]
@sickelap  commit hooks were failing because of some other files have fixed it mostly were because of unused imports

**Comment by derneuere** [comment]
Great work @AnkurPrabhu !!! 🎉

**Comment by AnkurPrabhu** [comment]
@derneuere hey just wanted to know is it working as expected ?

**Comment by derneuere** [comment]
@AnkurPrabhu I think so :) It will be part of the next release!
